Automattic/jetpack
# IDC: add detection for possibly dynamic HTTP_HOST in site URLs

Closes #14143

#### Changes proposed in this Pull Request:

* Adds a `detect_possible_dynamic_site_url` function which when an IDC has occurred, will try reading the WP Configuration file looking for potentially dynamic `HTTP_HOST` being used for the `WP_SITEURL` or `WP_HOME` definitions.
* When potentially dynamic URL usage is detected, a notice is shown on the IDCScreen component with a support link for more info. We say "potentially" because finding `HTTP_HOST` isn't fully indicative that the URL will be dynamic.
* New [support doc](https://jetpack.com/support/getting-started-with-jetpack/fixing-jetpack-connection-issues/resolving-jetpack-connection-issues-caused-by-dynamic-site-urls/) has been created. WCPay team [notified](https://github.com/Automattic/woocommerce-payments/issues/3825).

#### Jetpack product discussion

#14143

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

- Pull down changes to local JP dev env.
- `jetpack build packages/identity-crisis` to build things.
- Cause an IDC. Easiest to do with the Jetpack Debug Tools: turn on the IDC Simulator, then enable the IDC Simulation and store the options. Try refreshing a page (e.g. plugins, posts) to see the IDC screen; if you don't see it try editing a post to cause a sync.
- At this point a falsey value will have been stored for the `jetpack_idc_possible_dynamic_site_url_detected` transient. Delete this transient: `jetpack docker wp option delete _transient_jetpack_idc_possible_dynamic_site_url_detected`
- Now that you are seeing the IDC screen, try testing the new functionality. In your `wp-config.php` file, add the line `define('WP_SITEURL', 'http://' . $_SERVER['HTTP_HOST'] . '/');` and refresh the WP page you viewed with the IDC screen.
- This should hopefully have stored a `jetpack_idc_possible_dynamic_site_url_detected` transient with a truthy value. Feel free to remove the wp-config line you added since the transient value is stored/used for an hour.
- When transient value is true, we show a notice to admins as such (copy & support link WIP):

![Markup on 2022-02-14 at 13:23:27](https://user-images.githubusercontent.com/15803018/153940555-3cdfe2b3-9ed5-49d0-ba7a-50d46c5b2365.png)
